Content.Server/Administration/Managers/UsernameRuleManager.cs

```diff
@@ -0,0 +1,305 @@
+            && player is not null
+            && !_adminManager.HasAdminFlag(player, AdminFlags.Host)
+        )
+        {
+            return;
+        }
+
+        ClearCompiledRegex(restrictionId);
+
+        var adminName = removingAdmin == null
+            ? Loc.GetString("system-user")
+            : (await _db.GetPlayerRecordByUserId(removingAdmin.Value))?.LastSeenUserName ?? Loc.GetString("system-user");
+
+        var logMessage = Loc.GetString(
+            "server-username-rule-remove",
+            ("admin", adminName),
+            ("expression", rule.Expression),
+            ("message", rule.Message));
+
+        _sawmill.Info(logMessage);
+        _chatManager.SendAdminAlert(logMessage);
+
+        await _db.RemoveServerUsernameRuleAsync(restrictionId, removingAdmin, DateTimeOffset.Now);
+    }
+
+    public async Task<UsernameBanStatus> IsUsernameBannedAsync(string username)
+    {
+        var whitelist = await _db.CheckUsernameWhitelistAsync(username);
+        if (whitelist)
+        {
+            return new UsernameBanStatus("", false, false);
+        }
+
+        // check simple rules
+        if (_cachedUsernames.ContainsKey(username))
+        {
+            var rule = _cachedUsernameRules[_cachedUsernames[username]];
+            var fullMessage = ServerUsernameRuleDef.FormatUsernameViolationMessage(_cfgManager, _localizationManager, rule.Message);
+            return new UsernameBanStatus(fullMessage, rule.ExtendToBan, true);
+        }
+
+        // check regex rules
+        foreach (var rule in _cachedUsernameRules.Values)
+        {
+            if (rule.CompiledRule?.IsMatch(username) ?? false)
+            {
+                var fullMessage = ServerUsernameRuleDef.FormatUsernameViolationMessage(_cfgManager, _localizationManager, rule.Message);
+                return new UsernameBanStatus(fullMessage, rule.ExtendToBan, true);
+            }
+        }
+
+        return new UsernameBanStatus("", false, false);
+    }
+
+    public async void Restart()
+    {
+        var rules = await _db.GetServerUsernameRulesAsync(false);
+
+        if (rules == null)
+        {
+            _sawmill.Warning("service restart failed");
+            return;
+        }
+
+        _cachedUsernameRules.Clear();
+        _cachedUsernames.Clear();
+
+        foreach (var ruleDef in rules)
+        {
+            if (ruleDef.Id == null)
+            {
+                continue;
+            }
+
+            CacheCompiledRegex(ruleDef.Id ?? -1, ruleDef.Regex, ruleDef.Expression, ruleDef.Message, ruleDef.ExtendToBan);
+            KickMatchingConnectedPlayers(ruleDef.Id ?? -1, ruleDef, "username rule service restart");
+        }
+    }
+
+    private void KickMatchingConnectedPlayers(int id, ServerUsernameRuleDef def, string source)
+    {
+        if (!_cachedUsernameRules.ContainsKey(id))
+        {
+            return;
+        }
+
+        (Regex? compiledRule, string expression, _, _) = _cachedUsernameRules[id];
```

Comment:
This code does not feel like it's made easier to read by doing this deconstruction.